File: celesta/celesta-test/score/simpleCases/testSimpleCases.py
# coding=UTF-8

from java.sql import Timestamp
from java.time import LocalDateTime
from ru.curs.celesta.syscursors import LogCursor

from simpleCases._simpleCases_orm import getDateForViewCursor, viewWithGetDateCursor, zeroInsertCursor, duplicateCursor
from ru.curs.celesta.unit import TestClass, CelestaTestCase


def preInsert(logCursor):
    logCursor.tablename = "getDateForView"

def postInsert(logCursor):
    logCursor.sessionid = "1"

def preUpdate(logCursor):
    logCursor.tablename = "zeroInsert"

def postUpdate(logCursor):
    logCursor.sessionid = "2"

isPreDeleteDone = False
isPostDeleteDone = False

def resetFlags():
    global isPreDeleteDone
    global isPostDeleteDone
    isPreDeleteDone = False
    isPostDeleteDone = False

def preDelete(logCursor):
    global isPreDeleteDone
    isPreDeleteDone = True
    logCursor.tablename = "table2"

def postDelete(logCursor):
    global isPostDeleteDone
    isPostDeleteDone = True
    logCursor.sessionid = "2"


@TestClass
class TestSimpleCases(CelestaTestCase):
    def test_getdate_in_view(self):
        tableCursor = getDateForViewCursor(self.context)
        tableCursor.deleteAll()

        viewCursor = viewWithGetDateCursor(self.context)
        self.assertEquals(0, viewCursor.count())

        tableCursor.date = Timestamp.valueOf(LocalDateTime.now().minusDays(1))
        tableCursor.insert()
        self.assertEquals(0, viewCursor.count())

        tableCursor.clear()
        tableCursor.date = Timestamp.valueOf(LocalDateTime.now().plusDays(1))
        tableCursor.insert()
        self.assertEquals(1, viewCursor.count())


    def test_zero_insert(self):
        c = zeroInsertCursor(self.context)
        c.deleteAll()

        c.insert()
        print c.id
        print c.date


    def test_triggers_on_sys_cursors(self):
        resetFlags()
        c = LogCursor(self.context)

        LogCursor.onPreInsert(preInsert)
        LogCursor.onPostInsert(postInsert)
        LogCursor.onPreUpdate(preUpdate)
        LogCursor.onPostUpdate(postUpdate)
        LogCursor.onPreDelete(preDelete)
        LogCursor.onPostDelete(postDelete)

        c.userid = '1'
        c.sessionid = '0'
        c.grainid = 'simpleCases'
        c.tablename = "zeroInsert"
        c.action_type = "I"
        c.insert()

        self.assertEquals("getDateForView", c.tablename)
        self.assertEquals("1", c.sessionid)

        c.update()

        self.assertEquals("zeroInsert", c.tablename)
        self.assertEquals("2", c.sessionid)

        self.assertFalse(isPreDeleteDone)
        self.assertFalse(isPostDeleteDone)

        c.delete()

        self.assertTrue(isPreDeleteDone)
        self.assertTrue(isPostDeleteDone)
    
    def test_try_insert(self):
        c = duplicateCursor(self.context)
        c.deleteAll()
        c.id = 10
        self.assertTrue(c.tryInsert())
        self.assertFalse(c.tryInsert())
        c.id = 12
        self.assertTrue(c.tryInsert())
        
    def test_cursor_caching(self):
        
        zic1 = self.context.create(zeroInsertCursor)
        zic2 = self.context.create(zeroInsertCursor)
        self.assertEquals(zic1, zic2)
        
        zic1.close()
        self.assertTrue(zic2.isClosed())
        
        zic3 = self.context.create(zeroInsertCursor)
        self.assertNotEquals(zic1, zic3)

File: celesta-test/src/test/java/ru/curs/celesta/script/TestSimpleCases.java
package ru.curs.celesta.script;

import org.junit.jupiter.api.TestTemplate;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ru.curs.celesta.CallContext;
import ru.curs.celesta.CelestaException;
import ru.curs.celesta.syscursors.LogCursor;
import simpleCases.CustomSequence;
import simpleCases.DuplicateCursor;
import simpleCases.ForTriggersCursor;
import simpleCases.GetDateForViewCursor;
import simpleCases.SimpleTableCursor;
import simpleCases.UsesequenceCursor;
import simpleCases.ViewWithGetDateCursor;
import simpleCases.ZeroInsertCursor;

import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.util.Date;
import java.util.concurrent.atomic.AtomicBoolean;

import static org.junit.jupiter.api.Assertions.*;

public class TestSimpleCases implements ScriptTest {

    private static final Logger LOGGER = LoggerFactory.getLogger(TestSimpleCases.class);

    @TestTemplate
    void test_getdate_in_view(CallContext context) {
        GetDateForViewCursor tableCursor = new GetDateForViewCursor(context);
        tableCursor.deleteAll();

        ViewWithGetDateCursor viewCursor = new ViewWithGetDateCursor(context);
        assertEquals(0, viewCursor.count());

        tableCursor.setDate(Timestamp.valueOf(LocalDateTime.now().minusDays(1)));
        tableCursor.insert();
        assertEquals(0, viewCursor.count());

        tableCursor.clear();
        tableCursor.setDate(Timestamp.valueOf(LocalDateTime.now().plusDays(1)));
        tableCursor.insert();
        assertEquals(1, viewCursor.count());
    }

    @TestTemplate
    void test_zero_insert(CallContext context) {
        ZeroInsertCursor c = new ZeroInsertCursor(context);
        c.deleteAll();

        c.insert();
        LOGGER.info("{}", c.getId());
        LOGGER.info("{}", c.getDate());
    }

    @TestTemplate
    void testInsertWithSequence(CallContext context) {
        UsesequenceCursor c = new UsesequenceCursor(context);
        c.setId(100);
        c.insert();
        //A value is provided by the sequence
        assertTrue(c.getVal() > 0);
    }

    @TestTemplate
    void test_triggers_on_sys_cursors(CallContext context) {
        LogCursor c = new LogCursor(context);
        AtomicBoolean isPreDeleteDone = new AtomicBoolean(false);
        AtomicBoolean isPostDeleteDone = new AtomicBoolean(false);

        LogCursor.onPreInsert(context.getCelesta(), logCursor ->
                logCursor.setTablename("getDateForView"));
        LogCursor.onPostInsert(context.getCelesta(), logCursor ->
                logCursor.setSessionid("1"));
        LogCursor.onPreUpdate(context.getCelesta(), logCursor ->
                logCursor.setTablename("zeroInsert"));
        LogCursor.onPostUpdate(context.getCelesta(), logCursor ->
                logCursor.setSessionid("2"));
        LogCursor.onPreDelete(context.getCelesta(), logCursor -> {
            isPreDeleteDone.set(true);
            logCursor.setTablename("table2");
        });
        LogCursor.onPostDelete(context.getCelesta(), logCursor -> {
            isPostDeleteDone.set(true);
            logCursor.setSessionid("2");
        });

        c.setUserid("1");
        c.setSessionid("0");
        c.setGrainid("simpleCases");
        c.setTablename("zeroInsert");
        c.setActionType("I");
        c.insert();

        assertEquals("getDateForView", c.getTablename());
        assertEquals("1", c.getSessionid());

        c.update();

        assertEquals("zeroInsert", c.getTablename());
        assertEquals("2", c.getSessionid());

        assertFalse(isPreDeleteDone.get());
        assertFalse(isPostDeleteDone.get());

        c.delete();

        assertTrue(isPreDeleteDone.get());
        assertTrue(isPostDeleteDone.get());
    }

    @TestTemplate
    void test_triggers_on_gen_cursors(CallContext context) {
        ForTriggersCursor c = new ForTriggersCursor(context);
        AtomicBoolean isPreDeleteDone = new AtomicBoolean(false);
        AtomicBoolean isPostDeleteDone = new AtomicBoolean(false);

        ForTriggersCursor.onPreInsert(context.getCelesta(), forTriggersCursor -> {
            CustomSequence s = new CustomSequence(forTriggersCursor.callContext());
            forTriggersCursor.setId((int) s.nextValue());
        });
        ForTriggersCursor.onPostInsert(context.getCelesta(), forTriggersCursor -> forTriggersCursor.setVal(2));
        ForTriggersCursor.onPreUpdate(context.getCelesta(), forTriggersCursor -> forTriggersCursor.setVal(3));
        ForTriggersCursor.onPostUpdate(context.getCelesta(), forTriggersCursor -> {
            CustomSequence s = new CustomSequence(forTriggersCursor.callContext());
            forTriggersCursor.setId((int) s.nextValue());
        });
        ForTriggersCursor.onPreDelete(context.getCelesta(), forTriggersCursor -> isPreDeleteDone.set(true));
        ForTriggersCursor.onPostDelete(context.getCelesta(), forTriggersCursor -> isPostDeleteDone.set(true));

        c.insert();

        assertEquals(1, c.getId().intValue());
        assertEquals(2, c.getVal().intValue());

        c.update();

        assertEquals(2L, c.getId().intValue());
        assertEquals(3, c.getVal().intValue());

        assertFalse(isPreDeleteDone.get());
        assertFalse(isPostDeleteDone.get());

        c.setId(1);

        c.delete();

        assertTrue(isPreDeleteDone.get());
        assertTrue(isPostDeleteDone.get());
    }

    @TestTemplate
    void test_try_insert(CallContext context) {
        DuplicateCursor c = new DuplicateCursor(context);
        c.deleteAll();
        c.setId(10);
        assertTrue(c.tryInsert());
        assertFalse(c.tryInsert());
        c.setId(12);
        assertTrue(c.tryInsert());
    }

    @TestTemplate
    void test_cursor_getCurrentKeyValues(CallContext context) {
        SimpleTableCursor c = new SimpleTableCursor(context);
        c.deleteAll();
//        c.setId(1);
        c.setName("ONE");
        c.insert();

        c = new SimpleTableCursor(context);
        c = c.iterator().next();

        Object[] keyValues = c.getCurrentKeyValues();

        assertEquals(1, keyValues.length);
//        assertEquals(1, keyValues[0]);
        assertInstanceOf(Integer.class, keyValues[0]);
    }

    @TestTemplate
    void test_cursor_getCurrentValues(CallContext context) {
        SimpleTableCursor c = new SimpleTableCursor(context);
        c.deleteAll();
//        c.setId(1);
        c.setName("ONE");
        c.insert();

        c = new SimpleTableCursor(context);
        c = c.iterator().next();

        Object[] values = c.getCurrentValues();

        assertEquals(3, values.length);

//        assertEquals(1, values[0]);
        assertInstanceOf(Integer.class, values[0]);

        assertEquals("ONE", values[1]);
        assertInstanceOf(String.class, values[1]);
    }


    @TestTemplate
    void test_autoincremental_insertion(CallContext context) {
        GetDateForViewCursor tableCursor = new GetDateForViewCursor(context);
        tableCursor.deleteAll();
        for (int i = 0; i < 7; i++) {
            tableCursor.clear();
            tableCursor.setDate(new Date());
            tableCursor.insert();
        }
        assertEquals(7, tableCursor.count());
    }

    @TestTemplate
    void test_duplicate_insertion(CallContext context) {
        DuplicateCursor c = new DuplicateCursor(context);
        c.deleteAll();
        c.setId(7);
        c.insert();

        String message = assertThrows(CelestaException.class, c::insert).getMessage();
        assertEquals("Record duplicate [7] already exists", message);
    }

    @TestTemplate
    void test_update_nonexistent(CallContext context) {
        DuplicateCursor c = new DuplicateCursor(context);
        c.deleteAll();
        c.setId(42);
        c.setVal(42);

        String message = assertThrows(CelestaException.class, c::update).getMessage();
        assertEquals("Record duplicate [42] does not exist.", message);
    }
}
